thunder/executors/nvfuserex_impl.py

```diff
@@ -2732,6 +2737,268 @@ def embedding(
+
+    if any(x is not None for x in (weight, size_average, reduce)):
+        return False
+
+    if label_smoothing != 0.0:
+        return False
+
+    return True
+
+
+def cross_entropy_fwd_meta(
+    a: TensorLike,
+    /,
+    target: TensorLike,
+    weight: None | TensorLike = None,
+    size_average: None | Any = None,
+    ignore_index: int = -100,
+    reduce: None | Any = None,
+    reduction: str = "mean",
+    label_smoothing: float = 0.0,
+) -> tuple[TensorLike, TensorLike, TensorLike, TensorLike]:
+    losses: TensorLike
+    check(
+        reduction == "mean",
+        lambda: f"cross entropy expected reduction to be  'mean' but was given {reduction}",
+    )
+    losses = TensorLike(like=a, shape=())
+
+    max_log_sum_exp: TensorLike = TensorLike(like=target, dtype=dtypes.float32)
+    num_valid_indices: TensorLike = TensorLike(like=losses, shape=(), dtype=dtypes.float32)
+    a_max: TensorLike = TensorLike(like=target, dtype=dtypes.float32)
+    return losses, a_max, max_log_sum_exp, num_valid_indices
+
+
+def cross_entropy_fwd(
+    a: TensorLike,
+    /,
+    target: TensorLike,
+    weight: None | TensorLike = None,
+    size_average: None | Any = None,
+    ignore_index: int = -100,
+    reduce: None | Any = None,
+    reduction: str = "mean",
+    label_smoothing: float = 0.0,
+    *,
+    fd: FusionDefinition,
+    lc_to_nv_map: dict,
+) -> Any:
+    nv_a = getnv(a, fd, lc_to_nv_map)
+    nv_target = getnv(target, fd, lc_to_nv_map)
+    nv_ignore_index = getnv(ignore_index, fd, lc_to_nv_map)
+
+    zero_scalar = fd.define_scalar(0, dtype=torch_dtype_to_nvfuser_dtype(dtypes.to_torch_dtype(a.dtype)))
+
+    # modify the labels to account for ignore index and then do a
+    # gather/ take_along_axis on the input tensor
+    ne = fd.ops.ne(nv_target, nv_ignore_index)
+    where_0 = fd.ops.where(ne, nv_target, zero_scalar)
+    where = fd.ops.broadcast_in_dim(where_0, shape=[nv_target.shape()[-1], 1], broadcast_dims=[0])
+    gather0 = fd.ops.take_along_axis(nv_a, where, dim=1)
+    gather = fd.ops.reshape(gather0, new_shape=[nv_target.shape()[-1]])
+
+    # compute log (sum(exp(A - max(A, dim=1), dim=1))
+    max = fd.ops.max(nv_a, 1)
+    max2 = fd.ops.broadcast_in_dim(max, shape=[nv_target.shape()[-1], 1], broadcast_dims=[0])
+    sub = fd.ops.sub(nv_a, max2)
+    exp = fd.ops.exp(sub)
+    sum_1 = fd.ops.sum(exp, 1)
+    log = fd.ops.log(sum_1)
+
+    # compute S = POST_GATHER(A) - max(A, dim=1) - log(sum(exp(A - max(A, dim=1), dim=1))
+    gather_sub = fd.ops.sub(gather, max)
+    log_softmax_post_gather = fd.ops.sub(gather_sub, log)
+
+    # set the values for ignore index to 0
+    neg = fd.ops.neg(log_softmax_post_gather)
+    where_1 = fd.ops.where(ne, neg, zero_scalar)
+
+    # sum_2_cvt computes the number of valid indices
+    sum_2 = fd.ops.sum(ne)
+    sum_2_cvt = fd.ops.cast(sum_2, dtype=DataType.Float)
+
+    # compute the mean
+    sum_3 = fd.ops.sum(where_1)
+    div = fd.ops.div(sum_3, sum_2_cvt)
```

Comment:
Couldn't sum_2_cvt be zero if target is filled with -1s and ignore_index is -1? In pytorch this case returns NaN.

Reply:
That's a great point - let me see if I can add this

Reply:
I think for `fwd` it should still work.
`div = sum_3 / sum_2_cvt` 

If all entries in the target/labels are ignore index valued then both `sum_3` and `sum_2_cvt` will be zero and we'll return Nan. Please let me know if this is wrong.